cpp/src/binning/bin.cu

```diff
@@ -0,0 +1,222 @@
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <thrust/functional.h>
+#include <thrust/binary_search.h>
+#include <thrust/execution_policy.h>
+#include <cudf/detail/valid_if.cuh>
+#include <thrust/pair.h>
+
+namespace cudf {
+
+// Sentinel used to indicate that an input value should be placed in the null
+// bin.
+// NOTE: In theory if a user decided to specify 2^31 bins this would fail. We
+// could make this an error in Python, but that is such a crazy edge case...
+constexpr size_type NULL_VALUE{-1};
+
+namespace detail {
+namespace {
+
+template <typename T, typename StrictWeakOrderingLeft, typename StrictWeakOrderingRight>
+struct bin_finder
+{
+    bin_finder(
+            device_span<T> left_span,
+            device_span<T> right_span
+            )
+        : m_left_span(left_span), m_right_span(right_span)
+    {}
+
+    __device__ size_type operator()(thrust::pair<T, bool> input_value) const
+    {
+        // Immediately return sentinel for null inputs.
+        if (!input_value.second)
+            return NULL_VALUE;
+
+        T value = input_value.first;
+        auto bound = thrust::lower_bound(thrust::seq,
+                m_left_span.begin(), m_left_span.end(),
+                value,
+                m_left_comp);
+
+        // Exit early and return sentinel for values not within the interval.
+        if ((bound == m_left_span.begin()) || (bound == m_left_span.end()))
+            return NULL_VALUE;
+
+        // We must subtract 1 because lower bound returns the first index
+        // _greater than_ the value. This is safe because bound == m_left edges
+        // would already have triggered a NULL_VALUE return above, so there's
+        // no risk of negative values or wraparound for -1 here.
+        auto index = bound - m_left_span.begin() - 1;
+        return (m_right_comp(value, m_right_span[index])) ? index : NULL_VALUE;
+    }
+
+    device_span<T> m_left_span{};  // The range of data containing all left bin edges.
+    device_span<T> m_right_span{};  // The range of data containing all right bin edges.
+    // TODO: Can I implement these as static members rather than making an instance on construction?
+    StrictWeakOrderingLeft m_left_comp{}; // Comparator used for left edges.
+    StrictWeakOrderingRight m_right_comp{}; // Comparator used for left edges.
+};
+
+
+struct filter_null_sentinel
+{
+    __device__ bool operator()(size_type i)
+    {
+        return (i != NULL_VALUE);
+    }
+};
+
+/// Bin the input by the edges in left_edges and right_edges.
+template <typename T, typename StrictWeakOrderingLeft, typename StrictWeakOrderingRight, bool InputIsNullable>
+std::unique_ptr<column> bin(column_view const& input, 
+                            column_view const& left_edges,
+                            column_view const& right_edges,
+                            rmm::mr::device_memory_resource * mr)
+{
+    auto output = cudf::make_numeric_column(data_type(type_to_id<size_type>()), input.size());
```

Comment:
Please enforce consistency for the namespace prefix `cudf::`. Either you don't add it at all, or add it to everywhere else in your code, such as `cudf::data_type`, `cudf::type_to_id` etc.

Reply:
Thanks for the tip. In general rather than maintain consistency within this file I've tried to follow the most common usage of different identifiers seen in other parts of the code. So far as I can see, there is very minimal consistency in this fashion in other files, preference seems to be given to only explicitly namespace things that are not commonly used (e.g. `device_span` in this file) while using the naked identifiers for more common ones (such as `size_type` and `column`). I'll leave this thread open in case anyone else feels strongly about this, but for now I've updated my code to at least be internally consistent for a given type.